File: js-code/quick-sort.js
function partition(arr, start, length) {
  var pivotIdx = start;
  var pivotVal = arr[pivotIdx];

  for (var i = start + 1; i < start + length; i++) {
    val = arr[i];
    if (pivotVal < val) {
      //do nothing
    } else {
      arr[i] = arr[pivotIdx + 1];
      arr[pivotIdx + 1] = pivotVal;
      arr[pivotIdx] = val;

      pivotIdx += 1;
    }
  }

  return pivotIdx;
}

function qs (arr, start, len) {
  if (len < 2) {
    return arr;
  }
  // start = typeof start !== 'undefined' ? start : 0;
  // len = typeof len !== 'undefined' ? len : arr.length;
  if (typeof start !== 'number') {
    start = 0;
  }

  if (typeof len != 'number') {
    len = arr.length 
  }

  pivotIdx = partition(arr, start, len);
  // console.log("start is " + start);
  // console.log("pivotIdx is " + pivotIdx);
  // console.log("len is " + len);

  leftLength = pivotIdx - start;
  rightLength = len - leftLength - 1;

  qs(arr, start, leftLength);
  return qs(arr, pivotIdx + 1, rightLength);
}

console.log(qs([5, 0, 2, 3, 9, 9, 4, 1, 0, 2, 2, 9, 1, 4]));
// console.log(qs([5, 0, 2, 3, 4, 9, 1, 10]));
// console.log(partition([4, 1, 6, 1, 2, 3, 5], 0, 7));

File: lib/quick_sort.rb
class QuickSort

  def self.partition(arr, start, len)
    pivot_idx = start
    pivot_val = arr[pivot_idx] 

    ((start+1)...(start + len)).each do |idx|
      val = arr[idx]
      if pivot_val > val
        arr[idx] = arr[pivot_idx + 1]
        arr[pivot_idx + 1] = arr[pivot_idx]
        arr[pivot_idx] = val
      end

    end
    pivot_idx
  end

  def self.my_qs!(arr, start = 0, len = arr.length)
    return arr if len < 2
    pivot_idx = partition(arr, start, len)
    left_length = pivot_idx - start
    right_length = len - left_length - 1
    my_qs!(arr, start, left_length)
    my_qs!(arr, pivot_idx + 1, right_length)
  end
end

p QuickSort.my_qs!([3, 4, 1, 2, 3, 6, 7, 2, 1])
